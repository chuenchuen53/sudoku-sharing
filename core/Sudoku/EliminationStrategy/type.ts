import type { Candidates, Position, SudokuElement } from "../type";
import type { SudokuLine } from "../SudokuLine";

export interface Elimination extends Position {
  elements: SudokuElement[];
}

export interface Highlight {
  position: Position;
  candidates: Candidates;
  isSecondaryPosition?: boolean;
}

export interface EliminationData {
  eliminations: Elimination[];
  relatedLines: SudokuLine[];
  highlights: Highlight[];
}

export enum EliminationStrategyType {
  LOCKED_CANDIDATES = "LOCKED_CANDIDATES",
  NAKED_PAIRS = "NAKED_PAIRS",
  NAKED_TRIPLETS = "NAKED_TRIPLETS",
  NAKED_QUADS = "NAKED_QUADS",
  HIDDEN_PAIRS = "HIDDEN_PAIRS",
  HIDDEN_TRIPLETS = "HIDDEN_TRIPLETS",
  HIDDEN_QUADS = "HIDDEN_QUADS",
  X_WING = "X_WING",
  XY_WING = "XY_WING",
}
