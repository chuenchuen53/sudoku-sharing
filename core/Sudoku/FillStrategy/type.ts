import type { Position, PositionAndValue } from "../type";
import type { SudokuLine } from "../SudokuLine";

export enum FillStrategyType {
  UNIQUE_MISSING = "UNIQUE_MISSING",
  NAKED_SINGLE = "NAKED_SINGLE",
  HIDDEN_SINGLE = "HIDDEN_SINGLE",
}

export interface FillInputValueData extends PositionAndValue {
  relatedLine?: SudokuLine;
  highlightWholeCell?: true;
  secondaryRelatedLines?: SudokuLine[];
  secondaryHighlight?: Position[];
}
