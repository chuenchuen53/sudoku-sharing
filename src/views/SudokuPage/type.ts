import type { SudokuElementWithZero, CellWithIndex, InputValueData } from "@/Sudoku/type";

export interface Highlight {
  element: SudokuElementWithZero;
  cell: CellWithIndex[];
  candidate: InputValueData[];
  invalid: CellWithIndex[];
}
