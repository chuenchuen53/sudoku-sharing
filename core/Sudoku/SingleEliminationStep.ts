import EliminationStrategy from "./EliminationStrategy/EliminationStrategy";
import Sudoku from "./Sudoku";
import type { EliminationStep } from "./SudokuSolver";

export default class SingleEliminationStep {
  private constructor() {}

  public static singularizeSteps(step: EliminationStep): EliminationStep[] {
    const strategy = step.elimination.strategy;

    const result: EliminationStep[] = [];
    const initGrid = step.grid;
    const sudoku = Sudoku.sudokuFromGrid(initGrid);

    for (const x of step.elimination.data) {
      const removals = EliminationStrategy.removalsFromEliminationData([x]);

      result.push({
        grid: Sudoku.cloneGrid(sudoku.grid),
        elimination: {
          strategy,
          data: [x],
        },
      });
      if (sudoku.removeElementInCandidates(removals) === 0) result.pop();
    }

    return result;
  }
}
