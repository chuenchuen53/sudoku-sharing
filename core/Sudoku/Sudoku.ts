import ArrUtil from "../utils/ArrUtil";
import { VirtualLineType } from "./type";
import type { SudokuElement, Candidates, Cell, InputValueData, Grid, VirtualLine, InputClues, RowColumn, Position } from "./type";

export default class Sudoku {
  public grid: Grid;
  public numberOfClues: number;
  public invalidCells: Cell[];
  private rows: readonly VirtualLine[];
  private columns: readonly VirtualLine[];
  private boxes: readonly VirtualLine[];

  constructor(clues: InputClues) {
    this.grid = Sudoku.createGrid(clues);
    this.numberOfClues = this.grid.reduce((acc, row) => acc + row.reduce((acc, cell) => (cell.clue ? acc + 1 : acc), 0), 0);
    const nineLenArr = new Array(9).fill(true);
    this.rows = Object.freeze(this.grid.map((row) => row));
    this.columns = Object.freeze(nineLenArr.map((_, i) => this.grid.map((row) => row[i])));
    this.boxes = Object.freeze(
      nineLenArr.map((_, i) => {
        const firstRowIndex = Sudoku.boxFirstLineIndex(i, VirtualLineType.ROW);
        const firstColumnIndex = Sudoku.boxFirstLineIndex(i, VirtualLineType.COLUMN);
        return nineLenArr.map((_, j) => this.grid[firstRowIndex + Math.floor(j / 3)][firstColumnIndex + (j % 3)]);
      }),
    );
    this.invalidCells = Sudoku.invalidCells(this.grid);
  }

  get isValid(): boolean {
    return this.invalidCells.length === 0;
  }

  get solved(): boolean {
    this.invalidCells = Sudoku.invalidCells(this.grid);
    return this.isValid && this.grid.every((row) => row.every((cell) => cell.clue || cell.inputValue));
  }

  static createEmptyGrid(): Grid {
    return ArrUtil.create2DArray<Cell>(9, 9, (rowIndex, columnIndex) => ({ rowIndex, columnIndex }));
  }

  public static invalidCells(grid: Grid): Cell[] {
    const nineLenArr = new Array(9).fill(true);
    const rows: VirtualLine[] = nineLenArr.map((_, i) => grid[i]);
    const columns: VirtualLine[] = nineLenArr.map((_, i) => grid.map((row) => row[i]));
    const boxes: VirtualLine[] = nineLenArr.map((_, i) => {
      const firstRowIndex = Sudoku.boxFirstLineIndex(i, VirtualLineType.ROW);
      const firstColumnIndex = Sudoku.boxFirstLineIndex(i, VirtualLineType.COLUMN);
      return nineLenArr.map((_, j) => grid[firstRowIndex + Math.floor(j / 3)][firstColumnIndex + (j % 3)]);
    });

    const fn = (vl: VirtualLine[]) => vl.map((x) => Sudoku.duplicatedValueInVirtualLine(x, "inputValue")).flat();
    return Sudoku.removeDuplicatedPositions([...fn(rows), ...fn(columns), ...fn(boxes)]);
  }

  static sudokuFromGrid(grid: Grid): Sudoku {
    const clues: InputClues = grid.map((row) => row.map((cell) => cell.clue ?? "0"));
    const sudoku = new Sudoku(clues);
    for (let i = 0; i < grid.length; i++) {
      for (let j = 0; j < grid[i].length; j++) {
        const { inputValue, candidates } = grid[i][j];
        if (inputValue) {
          sudoku.setInputValue({ rowIndex: i, columnIndex: j, value: inputValue }, false);
        } else if (candidates) {
          sudoku.setCandidates(i, j, candidates);
        }
      }
    }
    return sudoku;
  }

  static boxFirstLineIndex(boxIndex: number, type: RowColumn): number {
    switch (type) {
      case VirtualLineType.ROW:
        return Math.floor(boxIndex / 3) * 3;
      case VirtualLineType.COLUMN:
        return (boxIndex % 3) * 3;
    }
  }

  static getBoxIndex(rowIndex: number, columnIndex: number): number {
    return Math.floor(rowIndex / 3) * 3 + Math.floor(columnIndex / 3);
  }

  static getRowColumnIndexFromBoxIndexAndCellIndex(boxIndex: number, cellIndex: number): { rowIndex: number; columnIndex: number } {
    return {
      rowIndex: Math.floor(boxIndex / 3) * 3 + Math.floor(cellIndex / 3),
      columnIndex: (boxIndex % 3) * 3 + (cellIndex % 3),
    };
  }

  static isSamePos(p1: Position, p2: Position): boolean {
    return p1.rowIndex === p2.rowIndex && p1.columnIndex === p2.columnIndex;
  }

  static virtualLinesIntersections(line1: VirtualLine, line2: VirtualLine): Cell[] {
    return line1.filter((x) => line2.some((y) => Sudoku.isSamePos(x, y)));
  }

  static missingValuesInVirtualLine(virtualLine: VirtualLine): Candidates {
    const missing = Sudoku.candidatesFactory(true);
    virtualLine.forEach((cell) => {
      if (cell.clue) {
        missing[cell.clue] = false;
      } else if (cell.inputValue) {
        missing[cell.inputValue] = false;
      }
    });
    return missing;
  }

  static removeDuplicatedPositions(positions: Position[]): Position[] {
    return positions.filter((cur, index, self) => index === self.findIndex((x) => Sudoku.isSamePos(x, cur)));
  }

  static removeDuplicatedInputValueData(data: InputValueData[]): InputValueData[] {
    return data.filter(
      (cur, index, self) =>
        index === self.findIndex((x) => x.rowIndex === cur.rowIndex && x.columnIndex === cur.columnIndex && x.value === cur.value),
    );
  }

  static candidatesFactory(defaultValue: boolean, elements?: SudokuElement[]): Candidates {
    if (!elements) {
      return {
        "1": defaultValue,
        "2": defaultValue,
        "3": defaultValue,
        "4": defaultValue,
        "5": defaultValue,
        "6": defaultValue,
        "7": defaultValue,
        "8": defaultValue,
        "9": defaultValue,
      };
    } else {
      const candidates: Candidates = {
        "1": !defaultValue,
        "2": !defaultValue,
        "3": !defaultValue,
        "4": !defaultValue,
        "5": !defaultValue,
        "6": !defaultValue,
        "7": !defaultValue,
        "8": !defaultValue,
        "9": !defaultValue,
      };
      for (const element of elements) {
        candidates[element] = defaultValue;
      }
      return candidates;
    }
  }

  static candidatesCount(candidates: Candidates): number {
    return Object.values(candidates).reduce((acc, cur) => (cur ? acc + 1 : acc), 0);
  }

  static duplicatedValueInVirtualLine(virtualLine: VirtualLine, key: Extract<keyof Cell, "clue" | "inputValue">): Cell[] {
    const duplicatedCells: Cell[] = [];
    const values = key === "clue" ? virtualLine.map((x) => x.clue) : virtualLine.map((x) => x.clue ?? x.inputValue);
    values.forEach((x, ix, arr) => x && arr.some((y, iy) => ix !== iy && x === y && duplicatedCells.push(virtualLine[ix])));
    return duplicatedCells;
  }

  static allElements(): SudokuElement[] {
    return ["1", "2", "3", "4", "5", "6", "7", "8", "9"];
  }

  static cloneGrid(grid: Grid): Grid {
    const clone = ArrUtil.create2DArray<Cell>(9, 9, (rowIndex, columnIndex) => ({ rowIndex, columnIndex }));
    for (let i = 0; i < grid.length; i++) {
      for (let j = 0; j < grid[i].length; j++) {
        const { rowIndex, columnIndex, clue, inputValue, candidates } = grid[i][j];
        clone[i][j] = {
          rowIndex,
          columnIndex,
        };
        if (clue) {
          clone[i][j].clue = clue;
        } else if (inputValue) {
          clone[i][j].inputValue = inputValue;
        } else if (candidates) {
          clone[i][j].candidates = { ...candidates };
        }
      }
    }
    return clone;
  }

  private static createGrid(clues: InputClues): Grid {
    if (clues.length !== 9 || clues.some((x) => x.length !== 9)) throw new Error("Invalid input clues");
    const grid: Grid = ArrUtil.create2DArray<Cell>(9, 9, (rowIndex, columnIndex) => ({ rowIndex, columnIndex }));
    clues.forEach((row, i) =>
      row.forEach((clue, j) => {
        if (clue !== "0") grid[i][j].clue = clue;
      }),
    );

    return grid;
  }

  getRow(rowIndex: number): VirtualLine {
    return this.rows[rowIndex];
  }

  getColumn(columnIndex: number): VirtualLine {
    return this.columns[columnIndex];
  }

  getBoxFromRowColumnIndex(rowIndex: number, columnIndex: number): VirtualLine {
    return this.boxes[Sudoku.getBoxIndex(rowIndex, columnIndex)];
  }

  getBoxFromBoxIndex(boxIndex: number): VirtualLine {
    const rowIndex = Sudoku.boxFirstLineIndex(boxIndex, VirtualLineType.ROW);
    const columnIndex = Sudoku.boxFirstLineIndex(boxIndex, VirtualLineType.COLUMN);
    return this.getBoxFromRowColumnIndex(rowIndex, columnIndex);
  }

  getAllRows(): VirtualLine[] {
    return this.rows.slice();
  }

  getAllColumns(): VirtualLine[] {
    return this.columns.slice();
  }

  getAllBoxes(): VirtualLine[] {
    return this.boxes.slice();
  }

  getAllVirtualLines(type: VirtualLineType): VirtualLine[] {
    switch (type) {
      case VirtualLineType.ROW:
        return this.getAllRows();
      case VirtualLineType.COLUMN:
        return this.getAllColumns();
      case VirtualLineType.BOX:
        return this.getAllBoxes();
    }
  }

  getAllRelatedBoxesInRowOrColumn(lineType: RowColumn, lineIndex: number): VirtualLine[] {
    switch (lineType) {
      case VirtualLineType.ROW:
        return [
          this.getBoxFromRowColumnIndex(lineIndex, 0),
          this.getBoxFromRowColumnIndex(lineIndex, 3),
          this.getBoxFromRowColumnIndex(lineIndex, 6),
        ];
      case VirtualLineType.COLUMN:
        return [
          this.getBoxFromRowColumnIndex(0, lineIndex),
          this.getBoxFromRowColumnIndex(3, lineIndex),
          this.getBoxFromRowColumnIndex(6, lineIndex),
        ];
    }
  }

  getAllRelatedRowsOrColumnsInBox(lineType: RowColumn, boxIndex: number): VirtualLine[] {
    const firstIndex = Sudoku.boxFirstLineIndex(boxIndex, lineType);
    switch (lineType) {
      case VirtualLineType.ROW:
        return [this.getRow(firstIndex), this.getRow(firstIndex + 1), this.getRow(firstIndex + 2)];
      case VirtualLineType.COLUMN:
        return [this.getColumn(firstIndex), this.getColumn(firstIndex + 1), this.getColumn(firstIndex + 2)];
    }
  }

  getAllRelatedCells(cell: Cell): Cell[] {
    const row = this.getRow(cell.rowIndex);
    const column = this.getColumn(cell.columnIndex);
    const box = this.getBoxFromRowColumnIndex(cell.rowIndex, cell.columnIndex);
    return [...row, ...column, ...box].filter(
      (x, index, arr) => !Sudoku.isSamePos(x, cell) && arr.findIndex((y) => Sudoku.isSamePos(x, y)) === index,
    );
  }

  setCandidates(rowIndex: number, columnIndex: number, candidates: Candidates): void {
    const cell = this.grid[rowIndex][columnIndex];
    if (!cell.candidates) {
      this.grid[rowIndex][columnIndex].candidates = { ...candidates };
    } else {
      for (const element of Sudoku.allElements()) {
        cell.candidates[element] = candidates[element];
      }
    }
  }

  setInputValue({ rowIndex, columnIndex, value }: InputValueData, updateValidateInfo: boolean): boolean {
    const cell = this.grid[rowIndex][columnIndex];
    if (cell.clue) return false;

    cell.inputValue = value;
    if (cell.candidates) delete cell.candidates;

    if (updateValidateInfo) this.validatePuzzle();

    return true;
  }

  setInputValues(data: InputValueData[]): void {
    const haveChanged: boolean[] = data.map((x) => this.setInputValue(x, false));
    if (haveChanged.filter((x) => x).length) this.validatePuzzle();
  }

  removeInputValue({ rowIndex, columnIndex }: { rowIndex: number; columnIndex: number }, update: boolean): void {
    if (this.grid[rowIndex][columnIndex].inputValue) {
      delete this.grid[rowIndex][columnIndex].inputValue;
    }

    if (update) this.validatePuzzle();
  }

  addElementInCandidates(inputValueDataArr: InputValueData[]): number {
    let count = 0;

    inputValueDataArr.forEach(({ rowIndex, columnIndex, value }) => {
      const cell = this.grid[rowIndex][columnIndex];

      if (cell.candidates) {
        if (!cell.candidates[value]) {
          cell.candidates[value] = true;
          count++;
        }
      } else {
        cell.candidates = Sudoku.candidatesFactory(true, [value]);
        count++;
      }
    });

    return count;
  }

  removeElementInCandidates(inputValueDataArr: InputValueData[]): number {
    let count = 0;

    inputValueDataArr.forEach(({ rowIndex, columnIndex, value }) => {
      const cell = this.grid[rowIndex][columnIndex];
      if (cell.candidates?.[value]) {
        cell.candidates[value] = false;
        count++;
      }
    });

    return count;
  }

  clearAllCandidates(): void {
    for (let i = 0; i < this.grid.length; i++) {
      for (let j = 0; j < this.grid[i].length; j++) {
        if (this.grid[i][j].candidates) delete this.grid[i][j].candidates;
      }
    }
  }

  validatePuzzle(): void {
    this.invalidCells = Sudoku.invalidCells(this.grid);
  }
}
