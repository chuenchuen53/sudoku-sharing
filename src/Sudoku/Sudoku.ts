import ArrUtil from "../utils/ArrUtil";
import { VirtualLineType } from "./type";
import type {
  SudokuElement,
  Candidates,
  CellWithIndex,
  InputValueData,
  Grid,
  VirtualLine,
  InputClues,
  RowColumn,
  CheckVirtualLineDuplicateResult,
  ValidateDetail,
  Cell,
} from "./type";

export default class Sudoku {
  public grid: Grid;
  public numberOfClues: number;
  public isValid: boolean;
  public validateDetail: ValidateDetail;

  constructor(clues: InputClues) {
    this.grid = Sudoku.createGrid(clues);
    this.numberOfClues = this.grid.reduce(
      (acc, row) => acc + row.reduce((acc, cell) => (cell.clue ? acc + 1 : acc), 0),
      0
    );
    const { isValid, validateDetail } = this.validatePuzzle("clue");
    this.isValid = isValid;
    this.validateDetail = validateDetail;
  }

  get solved(): boolean {
    const { isValid, validateDetail } = this.validatePuzzle("inputValue");
    this.isValid = isValid;
    this.validateDetail = validateDetail;
    return this.isValid && this.grid.every((row) => row.every((cell) => cell.clue || cell.inputValue));
  }

  static boxFirstLineIndex(boxIndex: number, type: RowColumn) {
    switch (type) {
      case VirtualLineType.ROW:
        return Math.floor(boxIndex / 3) * 3;
      case VirtualLineType.COLUMN:
        return (boxIndex % 3) * 3;
    }
  }

  static getBoxIndex(rowIndex: number, columnIndex: number) {
    return Math.floor(rowIndex / 3) * 3 + Math.floor(columnIndex / 3);
  }

  static getRowColumnIndexFromBoxIndexAndCellIndex(
    boxIndex: number,
    cellIndex: number
  ): { rowIndex: number; columnIndex: number } {
    return {
      rowIndex: Math.floor(boxIndex / 3) * 3 + Math.floor(cellIndex / 3),
      columnIndex: (boxIndex % 3) * 3 + (cellIndex % 3),
    };
  }

  static isSamePos(c1: CellWithIndex, c2: CellWithIndex): boolean {
    return c1.rowIndex === c2.rowIndex && c1.columnIndex === c2.columnIndex;
  }

  static virtualLinesIntersections(line1: VirtualLine, line2: VirtualLine): CellWithIndex[] {
    return line1.filter((x) => line2.some((y) => Sudoku.isSamePos(x, y)));
  }

  static missingInVirtualLine(virtualLine: VirtualLine): Candidates {
    const missing = Sudoku.candidatesFactory(true);
    virtualLine.forEach((cell) => {
      if (cell.clue) {
        missing[cell.clue] = false;
      } else if (cell.inputValue) {
        missing[cell.inputValue] = false;
      }
    });
    return missing;
  }

  static removeDuplicatedInputValueData(data: InputValueData[]) {
    return data.filter(
      (cur, index, self) =>
        index ===
        self.findIndex((x) => x.rowIndex === cur.rowIndex && x.columnIndex === cur.columnIndex && x.value === cur.value)
    );
  }

  static candidatesFactory(defaultValue: boolean, elements?: SudokuElement[]) {
    if (!elements) {
      return {
        "1": defaultValue,
        "2": defaultValue,
        "3": defaultValue,
        "4": defaultValue,
        "5": defaultValue,
        "6": defaultValue,
        "7": defaultValue,
        "8": defaultValue,
        "9": defaultValue,
      };
    } else {
      return {
        "1": elements.includes("1") ? defaultValue : !defaultValue,
        "2": elements.includes("2") ? defaultValue : !defaultValue,
        "3": elements.includes("3") ? defaultValue : !defaultValue,
        "4": elements.includes("4") ? defaultValue : !defaultValue,
        "5": elements.includes("5") ? defaultValue : !defaultValue,
        "6": elements.includes("6") ? defaultValue : !defaultValue,
        "7": elements.includes("7") ? defaultValue : !defaultValue,
        "8": elements.includes("8") ? defaultValue : !defaultValue,
        "9": elements.includes("9") ? defaultValue : !defaultValue,
      };
    }
  }

  static candidatesCount(candidates: Candidates): number {
    let count = 0;
    for (const key in candidates) {
      if (candidates[key as SudokuElement]) count++;
    }
    return count;
  }

  static checkVirtualLineHaveDuplicateValue(
    virtualLine: VirtualLine,
    key: Extract<keyof CellWithIndex, "clue" | "inputValue">
  ): CheckVirtualLineDuplicateResult {
    const duplicatedCells: CellWithIndex[] = [];
    const values = key === "clue" ? virtualLine.map((x) => x.clue) : virtualLine.map((x) => x.clue ?? x.inputValue);
    values.forEach(
      (x, ix, arr) => x && arr.some((y, iy) => ix !== iy && x === y && duplicatedCells.push(virtualLine[ix]))
    );
    const haveDuplicate = duplicatedCells.length > 0;
    return { haveDuplicate, duplicatedCells };
  }

  static allElements(): SudokuElement[] {
    return ["1", "2", "3", "4", "5", "6", "7", "8", "9"];
  }

  static loopCandidates(fn: (sudokuElement: SudokuElement) => void) {
    const allElements = Sudoku.allElements();
    for (let i = 0; i < allElements.length; i++) {
      fn(allElements[i]);
    }
  }

  private static createGrid(clues: InputClues): Grid {
    if (clues.length !== 9 || clues.some((x) => x.length !== 9)) throw new Error("Invalid input clues");
    const grid: Grid = ArrUtil.create2DArray<Cell>(9, 9, {});
    clues.forEach((row, i) =>
      row.forEach((clue, j) => {
        if (clue !== "0") grid[i][j].clue = clue;
      })
    );

    return grid;
  }

  getRow(rowIndex: number): VirtualLine {
    return this.grid[rowIndex].map((x, columnIndex) => ({
      ...x,
      rowIndex,
      columnIndex,
    }));
  }

  getColumn(columnIndex: number): VirtualLine {
    return this.grid.map((row, rowIndex) => ({
      ...row[columnIndex],
      rowIndex,
      columnIndex,
    }));
  }

  getBoxFromRowColumnIndex(rowIndex: number, columnIndex: number): VirtualLine {
    const box: VirtualLine = [];
    const firstRowIndex = rowIndex - (rowIndex % 3);
    const firstColumnIndex = columnIndex - (columnIndex % 3);
    for (let i = firstRowIndex; i < firstRowIndex + 3; i++) {
      for (let j = firstColumnIndex; j < firstColumnIndex + 3; j++) {
        box.push({
          ...this.grid[i][j],
          rowIndex: i,
          columnIndex: j,
        });
      }
    }
    return box;
  }

  getBoxFromBoxIndex(boxIndex: number) {
    const rowIndex = Sudoku.boxFirstLineIndex(boxIndex, VirtualLineType.ROW);
    const columnIndex = Sudoku.boxFirstLineIndex(boxIndex, VirtualLineType.COLUMN);
    return this.getBoxFromRowColumnIndex(rowIndex, columnIndex);
  }

  getAllRows(): VirtualLine[] {
    return this.grid.map((_, rowIndex) => this.getRow(rowIndex));
  }

  getAllColumns(): VirtualLine[] {
    return this.grid[0].map((_, columnIndex) => this.getColumn(columnIndex));
  }

  getAllBoxes(): VirtualLine[] {
    return Array.from({ length: 9 }, (_, i) => this.getBoxFromBoxIndex(i));
  }

  getAllVirtualLines(type: VirtualLineType): VirtualLine[] {
    switch (type) {
      case VirtualLineType.ROW:
        return this.getAllRows();
      case VirtualLineType.COLUMN:
        return this.getAllColumns();
      case VirtualLineType.BOX:
        return this.getAllBoxes();
    }
  }

  getAllRelatedBoxesInRowOrColumn(lineType: RowColumn, lineIndex: number): VirtualLine[] {
    const boxes: VirtualLine[] = [];
    for (let i = 0; i < 9; i += 3) {
      lineType === VirtualLineType.ROW
        ? boxes.push(this.getBoxFromRowColumnIndex(lineIndex, i))
        : boxes.push(this.getBoxFromRowColumnIndex(i, lineIndex));
    }
    return boxes;
  }

  getAllRelatedRowsOrColumnsInBox(lineType: RowColumn, boxIndex: number): VirtualLine[] {
    const firstIndex = Sudoku.boxFirstLineIndex(boxIndex, lineType);
    const getLine = lineType === VirtualLineType.ROW ? this.getRow.bind(this) : this.getColumn.bind(this);
    return [getLine(firstIndex), getLine(firstIndex + 1), getLine(firstIndex + 2)];
  }

  getAllRelatedCells(cell: CellWithIndex): CellWithIndex[] {
    const row = this.getRow(cell.rowIndex);
    const column = this.getColumn(cell.columnIndex);
    const box = this.getBoxFromRowColumnIndex(cell.rowIndex, cell.columnIndex);
    return [...row, ...column, ...box].filter(
      (x, index, arr) => !Sudoku.isSamePos(x, cell) && arr.findIndex((y) => Sudoku.isSamePos(x, y)) === index
    );
  }

  setCandidates(rowIndex: number, columnIndex: number, candidates: Candidates) {
    this.grid[rowIndex][columnIndex].candidates = { ...candidates };
  }

  setInputValue({ rowIndex, columnIndex, value }: InputValueData, update: boolean) {
    if (this.grid[rowIndex][columnIndex].clue) {
      console.error("Cannot set input value to a cell with a clue");
      return;
    }

    this.grid[rowIndex][columnIndex].inputValue = value;
    delete this.grid[rowIndex][columnIndex].candidates;

    if (update) {
      const { isValid, validateDetail } = this.validatePuzzle("inputValue");
      this.isValid = isValid;
      this.validateDetail = validateDetail;
    }
  }

  setInputValues(data: InputValueData[]) {
    data.forEach((x) => this.setInputValue(x, false));
    if (data.length) {
      const { isValid, validateDetail } = this.validatePuzzle("inputValue");
      this.isValid = isValid;
      this.validateDetail = validateDetail;
    }
  }

  removeInputValue({ rowIndex, columnIndex }: { rowIndex: number; columnIndex: number }, update: boolean) {
    if (this.grid[rowIndex][columnIndex].inputValue) {
      delete this.grid[rowIndex][columnIndex].inputValue;
    }

    if (update) {
      const { isValid, validateDetail } = this.validatePuzzle("inputValue");
      this.isValid = isValid;
      this.validateDetail = validateDetail;
    }
  }

  clearAllCandidates() {
    for (let i = 0; i < this.grid.length; i++) {
      for (let j = 0; j < this.grid[i].length; j++) {
        delete this.grid[i][j].candidates;
      }
    }
  }

  validatePuzzle(type: Extract<keyof CellWithIndex, "clue" | "inputValue">): {
    isValid: boolean;
    validateDetail: ValidateDetail;
  } {
    const allRows = this.getAllRows();
    const allColumns = this.getAllColumns();
    const allBoxes = this.getAllBoxes();

    const rowDetail = allRows.map((x) => Sudoku.checkVirtualLineHaveDuplicateValue(x, type));
    const columnDetail = allColumns.map((x) => Sudoku.checkVirtualLineHaveDuplicateValue(x, type));
    const boxDetail = allBoxes.map((x) => Sudoku.checkVirtualLineHaveDuplicateValue(x, type));
    const validateDetail = {
      [VirtualLineType.ROW]: rowDetail,
      [VirtualLineType.COLUMN]: columnDetail,
      [VirtualLineType.BOX]: boxDetail,
    };
    const isValid =
      !rowDetail.some((x) => x.haveDuplicate) &&
      !columnDetail.some((x) => x.haveDuplicate) &&
      !boxDetail.some((x) => x.haveDuplicate);

    return { isValid, validateDetail };
  }
}
