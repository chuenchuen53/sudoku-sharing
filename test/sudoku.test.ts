import { assert, expect, describe, it } from "vitest";
import Sudoku from "../src/Sudoku";
import type { InputClues } from "../src/sudoku/type";

export const emptyPuzzle: InputClues = [
  ["0", "0", "0", "0", "0", "0", "0", "0", "0"],
  ["0", "0", "0", "0", "0", "0", "0", "0", "0"],
  ["0", "0", "0", "0", "0", "0", "0", "0", "0"],
  ["0", "0", "0", "0", "0", "0", "0", "0", "0"],
  ["0", "0", "0", "0", "0", "0", "0", "0", "0"],
  ["0", "0", "0", "0", "0", "0", "0", "0", "0"],
  ["0", "0", "0", "0", "0", "0", "0", "0", "0"],
  ["0", "0", "0", "0", "0", "0", "0", "0", "0"],
  ["0", "0", "0", "0", "0", "0", "0", "0", "0"],
];

export const testingPuzzle1: InputClues = [
  ["2", "0", "0", "0", "0", "0", "8", "6", "0"],
  ["0", "0", "0", "0", "4", "2", "0", "0", "0"],
  ["0", "1", "0", "0", "6", "0", "0", "4", "7"],
  ["3", "4", "5", "0", "2", "0", "0", "0", "1"],
  ["7", "2", "0", "0", "0", "0", "4", "0", "9"],
  ["8", "0", "0", "0", "0", "0", "5", "0", "6"],
  ["0", "0", "2", "0", "3", "0", "0", "0", "0"],
  ["0", "0", "0", "6", "8", "0", "0", "1", "2"],
  ["5", "0", "8", "0", "0", "0", "0", "0", "4"],
];

export const testingPuzzle2: InputClues = [
  ["0", "3", "0", "9", "0", "0", "0", "0", "0"],
  ["6", "0", "0", "2", "0", "0", "8", "0", "0"],
  ["8", "0", "0", "6", "1", "0", "5", "4", "9"],
  ["0", "0", "0", "0", "3", "2", "1", "0", "0"],
  ["2", "0", "8", "0", "4", "0", "0", "0", "0"],
  ["0", "0", "3", "1", "9", "0", "0", "0", "4"],
  ["9", "0", "2", "0", "0", "0", "0", "0", "5"],
  ["1", "0", "0", "0", "0", "0", "0", "6", "0"],
  ["0", "0", "4", "0", "6", "0", "9", "0", "8"],
];

describe("sudoku basic", () => {
  it("createPuzzle", () => {
    const sudoku = new Sudoku(emptyPuzzle);
    const grid = sudoku.createPuzzle(testingPuzzle1);
    const gridExpected = [
      [{ clue: "2" }, {}, {}, {}, {}, {}, { clue: "8" }, { clue: "6" }, {}],
      [{}, {}, {}, {}, { clue: "4" }, { clue: "2" }, {}, {}, {}],
      [{}, { clue: "1" }, {}, {}, { clue: "6" }, {}, {}, { clue: "4" }, { clue: "7" }],
      [{ clue: "3" }, { clue: "4" }, { clue: "5" }, {}, { clue: "2" }, {}, {}, {}, { clue: "1" }],
      [{ clue: "7" }, { clue: "2" }, {}, {}, {}, {}, { clue: "4" }, {}, { clue: "9" }],
      [{ clue: "8" }, {}, {}, {}, {}, {}, { clue: "5" }, {}, { clue: "6" }],
      [{}, {}, { clue: "2" }, {}, { clue: "3" }, {}, {}, {}, {}],
      [{}, {}, {}, { clue: "6" }, { clue: "8" }, {}, {}, { clue: "1" }, { clue: "2" }],
      [{ clue: "5" }, {}, { clue: "8" }, {}, {}, {}, {}, {}, { clue: "4" }],
    ];
    expect(grid).toEqual(gridExpected);
  });

  it("createPuzzle", () => {
    const sudoku = new Sudoku(emptyPuzzle);
    const grid = sudoku.createPuzzle(testingPuzzle2);
    const gridExpected = [
      [{}, { clue: "3" }, {}, { clue: "9" }, {}, {}, {}, {}, {}],
      [{ clue: "6" }, {}, {}, { clue: "2" }, {}, {}, { clue: "8" }, {}, {}],
      [{ clue: "8" }, {}, {}, { clue: "6" }, { clue: "1" }, {}, { clue: "5" }, { clue: "4" }, { clue: "9" }],
      [{}, {}, {}, {}, { clue: "3" }, { clue: "2" }, { clue: "1" }, {}, {}],
      [{ clue: "2" }, {}, { clue: "8" }, {}, { clue: "4" }, {}, {}, {}, {}],
      [{}, {}, { clue: "3" }, { clue: "1" }, { clue: "9" }, {}, {}, {}, { clue: "4" }],
      [{ clue: "9" }, {}, { clue: "2" }, {}, {}, {}, {}, {}, { clue: "5" }],
      [{ clue: "1" }, {}, {}, {}, {}, {}, {}, { clue: "6" }, {}],
      [{}, {}, { clue: "4" }, {}, { clue: "6" }, {}, { clue: "9" }, {}, { clue: "8" }],
    ];
    expect(grid).toEqual(gridExpected);
  });
});
