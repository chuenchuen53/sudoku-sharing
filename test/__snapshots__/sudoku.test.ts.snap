// Vitest Snapshot v1

exports[`sudoku basic > snapshot 1`] = `
{
  "foo": "bar",
}
`;

exports[`suite name > snapshot 1`] = `
{
  "foo": "bar",
}
`;
